spock-core/src/main/java/org/spockframework/runtime/model/DataVariableMultiplication.java

```diff
@@ -0,0 +1,41 @@
+/*
```

Comment:
Imho every file should have a copyright header in some way, especially in an OSS project. This way the license of the content is clear even if only this single file appears somewhere standalone. I did a quick analysis over all `java` and `groovy` files in current Spock `master` and yeah, about 25% miss the copyright header. So by majority decision of the source itself, they should be there. :-D
It's just very easy to forget them, and also forgetting to mention in the review, as you see with the mock maker PR where I didn't complain about it. :-(
So I would say yes, please also add them to the files you created in your PRs. :-)

Reply:
@Vampire, and IMO this is just white noise und legally totally unnecessary. One licence and copyright file in the root folder of the project and in each public artifact on Maven Central should be enough. This whole ceremony of copying and pasting redundant text data on top ofevery text and source file has zero value. It is just as easy to remove or replace it as to add it. It is not like this is some kind of digital watermark or so. I simply find rules like this one ridiculous, excuse my French.

Reply:
Again, this is not to protect the file, if one has the file they can always use it in illegal ways.
It is to make it crystal clear under which license a file's content is distributed.
If you just send one file standalone to someone, then if it does not have a license header, it is simply unclear under which license the code can be used, and thus the code cannot be used at all, unless the receiver of that file finds out which project the file is from and in exactly the state he got the file, because even if it is obvious that the file is for example from the Spock project, one would still need to find out whether any of the commits contains the file in exactly that state, otherwise there could be modifications done to the file that are covered by a different license.
Having the license header just makes it clear under which license the code in the file is distributed.

This is also how the creators of the license recommend to use the license, and if in doubt, I'd follow their advice, as they are the experts that were also competent enough to write the license. :-)
Quote from https://www.apache.org/legal/apply-license.html:
> Each original source document (code and documentation, but not the LICENSE and NOTICE files) should include [a short license header](https://www.apache.org/legal/src-headers.html#headers) at the top.

Reply:
Thanks for the advice, but I see no reason to watch my tone, because I did not attack you or anyone else personally. I am criticising a state of affairs in the FOSS community. I am taking the liberty to call something a pseudo reason, if I think it is one. As for my definition of insanity in this case, I was not referring to Einstein at all. Thanks also for the hint to use Spotless, but my whole point is that it is unnecessary, and I gave reasons for that, if you recall.

I am looking at this issue both legally and from a time economics and management perspective: Legally, I think a FOSS project is safe, if the licence is kept in a single, easily discoverable place, which in all FOSS projects I know it is. Specifically at GitHub, we can even make the licence even more easily discoverable by putting it in the project's meta data and displaying them prominently in the project header:

![image](https://github.com/spockframework/spock/assets/1537384/bbdcd365-6cdc-4057-8c8e-03c6e7660508)

I would be super relaxed and confident to win any legal case, if I kept any FOSS licence in, say, a LICENSE file in the top level directory and in any public artifact (JAR), if I or my lawyer had to argue such a case in front of any court against someone claiming that because I did not put copyright headers into some or even all of my source code files, said licence would not apply to them. The important question is: Could the other party have known which licence applies to the code? Yes, they could, and effortlessly so. Case dismissed. All this historically handed-down overhead from times before the existence or widely accepted use of infrastructure like SCM systems available online when people discussed code by sending it via e-mail instead of sending links is simply no longer necessary. Anyone who would try to argue that the project licence did not apply to single files missing headers would get the crap beaten out of them in court, and the FOSS project would win with flying colours.

Reply:
> Thanks for the advice, but I see no reason to watch my tone

That's exactly part of the problem and why others need to tell you.
If someone tells you something like that, you obviously **should** do so.

> because I did not attack you or anyone else personally.

Yes, you did. I acknowledge, that you don't mean what you say the way you say it.
But you told that the reasons I list are pseudo reasons, which **is** a personal attack.
Jut because you disagree out of pure personal liking, does not make the reasons "pseudo".
And you said that it is "insanity to work like this", so another personal attack on my and others' work.

>  I am looking at this issue both legally and from a time economics and management perspective: Legally, I think a FOSS project is safe, if the licence is kept in a single, easily discoverable place, which in all FOSS projects I know it is. Specifically at GitHub, we can even make the licence even more easily discoverable by putting it in the project's meta data and displaying it prominently in the project header:

It's really sad I have to repeat it again. Especially if you try to come with legal reasons, one would think you read properly before you answer. You are even legally save if you do not put any mention of any licence anywhere. You always keep the copyright on the code you wrote. So even then you **are** legally safe. That was not the point at all though. The point of specifying a license is to give other people the right to use your code under the terms you specified. And having the license headers in the files just makes it crystal clear under which license a file is distributed, even if it is sent around standalone, which I often did already, even if you do not.

But please let's stop this pointless discussion here.
You have your opinion that obviously will not change, and I strongly disagree, especially as you do not even refer to what I am saying.
If you are really keen on getting those thrown out, please try to convince Leonard outside a concrete PR, or in a PR that throws them out and if he agrees with you and decides for this project to not follow best practices, but to remove the license headers, I have no problem with that, though I would of course voice my concerns.

Reply:
Believe it or not, @Vampire: I mean exactly what I say, but still I am **not attacking anyone personally but criticising common practices** which I believe have little to no value. If I call an argument a pseudo argument, it is because I think it is unfounded or the arguments backing it are very weak. If I call a practice insanity, it is because I think it creates lots of waste and no tangible value. I allow you to hold a different opinion and defend it. Please grant me the same favour. If you do not agree with my choice of language, sorry for that. Deal with it.